### PR TITLE
#223/create a get endpoint to get the messages from a chat room

### LINKED ISSUE #221: Create a database method to add a chat room message
A part of the user story: _As a chat room member, I want to be able to send messages and receive messages from other members in real time_

### LINKED ISSUE #222: Create a database method to get the messages from a chat room
A part of the user story: _As a chat room member, I want to be able to send messages and receive messages from other members in real time_

### LINKED ISSUE #223: Create a GET endpoint to get the messages from a chat room
A part of the user story: _As a chat room member, I want to be able to send messages and receive messages from other members in real time_

Details: Create a GET /chat/room/{roomId}/messages endpoint to get the messages from a chat room. _This endpoint must be protected_

### PR DESCRIPTION
**Pull request by gbudau**
close #221
close #222
close #223

### CODE REVIEW

File: transcendence-app/src/chat/chat.controller.ts [hunk 1]
```diff
@@ -130,10 +133,44 @@ export class ChatController {
...
+  ): Promise<ChatRoomMessageWithUser[] | null> {
+    const chatroom = await this.chatService.getChatroomById(chatroomId);
+    if (!chatroom) {
+      throw new NotFoundException();
+    }
+    const isChatMember = await this.chatMemberService.getById(
+      chatroomId,
+      user.id,
+    );
+    if (!isChatMember) {
```
Comment (Damrod) [inline transcendence-app/src/chat/chat.controller.ts:164]: maybe this can also be in L 122?

**Comment by Damrod** [comment]
It's a bit hard to test this without first seeding the chatroommessage table. How do you do it?

File: transcendence-app/src/shared/dtos/pagination-query.dto.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
+import { ApiPropertyOptional } from '@nestjs/swagger';
```
Comment (Damrod) [inline transcendence-app/src/shared/dtos/pagination-query.dto.ts:1]: Thank you for this feature!

File: transcendence-app/src/chat/chat.controller.ts (hunk 1 above)
Comment (gbudau) [reply transcendence-app/src/chat/chat.controller.ts:168]: :+1: done

File: transcendence-app/src/chat/infrastructure/db/chatmember.repository.ts [hunk 3]
```diff
@@ -1,7 +1,12 @@
-import { ChatMember, ChatMemberWithUser } from '../../chatmember.domain';
+import { ChatMemberWithUser } from '../../chatmember.domain';
+import { ChatMemberEntity } from './chatmember.entity';
 
 export abstract class IChatMemberRepository {
-  abstract add(chatMember: ChatMember): Promise<ChatMember | null>;
+  abstract add(chatMember: ChatMemberEntity): Promise<ChatMemberEntity | null>;
```
Comment (Damrod) [inline transcendence-app/src/chat/infrastructure/db/chatmember.repository.ts:5]: In user, the add function takes a User (from domain) and returns a Promise<User|null>. Why here you change this to use entity instead?

**Comment by gbudau** [comment]
What I did was a bit of manual setup.

1. Create a chatroom.
2. Login in with one user and join the chatroom by making a request to http://localhost:3000/api#/chat/ChatController_createChatRoomMember with the chatroom ID.
3. Login in with another user and join the same chatroom 
4. Insert chatroom messages into the database using the chatroom ID and user IDs.

```sql
INSERT INTO
  ChatRoomMessage ("chatId", "userId", "content")
VALUES
  (
    'your-chatroom-id',
    'your-user-id',
    'message content'
  );
```

File: transcendence-app/src/chat/infrastructure/db/chatmember.repository.ts (hunk 3 above)
Comment (gbudau) [reply transcendence-app/src/chat/infrastructure/db/chatmember.repository.ts:5]: We might want to make some changes to our repository methods.
In the current implementation, we return an object with the same properties as the entity/domain. They seem interchangeable.
We could return an entity instance, which we could serialize using `ClassSerializerInterceptor`.

Comment (gbudau) [reply transcendence-app/src/chat/infrastructure/db/chatmember.repository.ts:5]: I've created a new task for this #301

**Review by fizcamposancos — APPROVED** [review]
LGTM!
One general question, since the controller methods are getting bigger, calling the service, doing some logic and then throwing exceptions or the response. What is better, a simpler controller and the logic in the service or a simpler service and the controller just catching error or the response?

**Comment by gbudau** [comment]
I don't know. I prefer to throw the HTTP exceptions from the controllers when I can because it's easier to document the endpoints, and maybe we want to use the services with other protocols, like WebSocket.
But I think the logic should be in the service as much as possible.

**Comment by gbudau** [comment]
Thanks!

File: transcendence-app/src/chat/infrastructure/db/chatmember.repository.ts (hunk 3 above)
Comment (gbudau) [reply transcendence-app/src/chat/infrastructure/db/chatmember.repository.ts]: I made a POC with repositories that return entities instances and no domains #303

**Pull request merged**